flink-yarn-tests/src/test/java/org/apache/flink/yarn/YARNITCase.java

```diff
@@ -133,6 +140,38 @@ public void testPerJobWithArchive() throws Exception {
         runTest(() -> deployPerJob(flinkConfig, archiveJobGraph, true));
     }
 
+    @Test
+    public void testPerJobModeWithAsyncClusterClientProvider() throws Exception {
+        runTest(
+                () -> {
+                    Configuration configuration =
+                            createDefaultConfiguration(YarnConfigOptions.UserJarInclusion.FIRST);
+                    JobGraph jobGraph = getTestingJobGraph();
+                    jobGraph.setJobType(JobType.STREAMING);
+                    try (final YarnClusterDescriptor yarnClusterDescriptor =
+                            createYarnClusterDescriptor(configuration)) {
+
+                        ClusterClientProvider<ApplicationId> clusterClientProvider =
+                                deployJobCluster(yarnClusterDescriptor, jobGraph);
+
+                        ApplicationId applicationId = yarnClusterDescriptor.getApplicationId();
+                        YarnClient yarnClient = getYarnClient();
+                        ApplicationReport report = yarnClient.getApplicationReport(applicationId);
+
+                        assertEquals(ACCEPTED, report.getYarnApplicationState());
+
+                        ClusterClient<ApplicationId> clusterClient =
+                                clusterClientProvider.getClusterClient();
+
+                        report = yarnClient.getApplicationReport(applicationId);
+                        assertEquals(RUNNING, report.getYarnApplicationState());
+
+                        checkApplicationFinished(
+                                clusterClient, yarnClusterDescriptor, jobGraph, configuration);
+                    }
+                });
```

Comment:
What exactly is this test testing that we aren't testing yet?

Reply:
Just verify that once application is accepted, the cluster client provider will be returned.
And it will block waiting app state from accepted to running utill invoking `getClusterClient` method.